scrapper.py

```diff
@@ -31,67 +55,291 @@ class Crawler:
+        Extracts news urls from the seed page
+        """
+        news = article_bs.select("div.col-md-9 div.col-md-6 h2 a")
+        links = [article.attrs["href"] for article in news]
+
+        return links
+
+    def _process_seed(self, page: str):
+        """
+        Extracts specified number of links from the seed page, filters and returns them
+        """
+        soup = BeautifulSoup(page, "lxml")
+
+        seed_articles = self._extract_url(soup)[: self.max_articles_per_seed]
+        rus_links = [x for x in seed_articles if re.search(r"[a-z-]+/$", x)]
+        links = rus_links[: self.max_articles - len(self.urls)]
+
+        return links
+
+
+class CrawlerRecursive(Crawler):
+    """
+    Recursive Crawler
+    """
+
+    def find_articles(self):
+        """
+        Finds articles
+        """
+        self._load_state()
+
+        seed, self.seed_urls[0] = self.get_search_urls()
+        page = fetch_page(seed)
+        links = self._process_seed(page)
+        self.urls.update(links)
+
+        if len(self.urls) == self.max_articles:
+            return None
+
+        self._save_state()
+
+        log.info("Seed page '%s' is processed.", seed)
+
+        time.sleep(random.uniform(1, 3))
+
+        return self.find_articles()
+
+    def get_search_urls(self) -> Tuple[str, str]:
+        """
+        Gets current and next seed pages
+        """
+        seed = self.seed_urls[0]
+        page = fetch_page(seed)
+
+        soup = BeautifulSoup(page, "lxml")
+        next_page = soup.select_one("ul.pagination li:has(span.current) + li a")
+
+        return seed, next_page["href"]
+
+    def _save_state(self) -> None:
+        with open(CRAWLER_STATE, "wb") as file:
+            pickle.dump({"urls": self.urls, "seed": self.seed_urls}, file)
+
+    def _load_state(self) -> None:
+        if os.path.exists(CRAWLER_STATE):
+            with open(CRAWLER_STATE, "rb") as file:
+                status: Dict[str, Any] = pickle.load(file)
+
+            self.seed_urls, self.urls = status["seed"], status["urls"]
 
 
 class ArticleParser:
     """
     ArticleParser implementation
     """
-    def __init__(self, full_url: str, article_id: int):
-        pass
 
-    def _fill_article_with_text(self, article_soup):
-        pass
+    article: Article
+    full_url: str
+    article_id: int
 
-    def _fill_article_with_meta_information(self, article_soup):
-        pass
+    def __init__(self, full_url: str, article_id: int):
+        self.full_url = full_url
+        self.article_id = article_id
+        self.article = Article(full_url, article_id)
 
     @staticmethod
-    def unify_date_format(date_str):
+    def unify_date_format(date_str: str) -> datetime:
         """
         Unifies date format
         """
-        pass
+        return datetime.fromisoformat(date_str)
+
+    @property
+    def processed(self) -> bool:
+        return self.full_url in self._load_state()
+
+    @property
+    def is_russian(self) -> bool:
```

Comment:
where is it used?

Reply:
nowhere actually, there was a different check for this; removed